airflow/providers/amazon/aws/triggers/ecs.py

```diff
@@ -0,0 +1,204 @@
+# specific language governing permissions and limitations
+# under the License.
+
+from __future__ import annotations
+
+import asyncio
+from typing import Any, AsyncIterator
+
+from botocore.exceptions import ClientError, WaiterError
+
+from airflow import AirflowException
+from airflow.providers.amazon.aws.hooks.ecs import EcsHook, EcsTaskLogFetcher
+from airflow.providers.amazon.aws.hooks.logs import AwsLogsHook
+from airflow.triggers.base import BaseTrigger, TriggerEvent
+
+
+class ClusterWaiterTrigger(BaseTrigger):
+    """
+    Polls the status of a cluster using a given waiter. Can be used to poll for an active or inactive cluster.
+
+    :param waiter_name: Name of the waiter to use, for instance 'cluster_active' or 'cluster_inactive'
+    :param cluster_arn: ARN of the cluster to watch.
+    :param waiter_delay: The amount of time in seconds to wait between attempts.
+    :param waiter_max_attempts: The number of times to ping for status.
+        Will fail after that many unsuccessful attempts.
+    :param aws_conn_id: The Airflow connection used for AWS credentials.
+    :param region: The AWS region where the cluster is located.
+    """
+
+    def __init__(
+        self,
+        waiter_name: str,
+        cluster_arn: str,
+        waiter_delay: int | None,
+        waiter_max_attempts: int | None,
+        aws_conn_id: str | None,
+        region: str | None,
+    ):
+        self.cluster_arn = cluster_arn
+        self.waiter_name = waiter_name
+        self.waiter_delay = waiter_delay if waiter_delay is not None else 15  # written like this to allow 0
+        self.attempts = waiter_max_attempts or 999999999
+        self.aws_conn_id = aws_conn_id
+        self.region = region
+
+    def serialize(self) -> tuple[str, dict[str, Any]]:
+        return (
+            self.__class__.__module__ + "." + self.__class__.__qualname__,
+            {
+                "waiter_name": self.waiter_name,
+                "cluster_arn": self.cluster_arn,
+                "waiter_delay": self.waiter_delay,
+                "waiter_max_attempts": self.attempts,
+                "aws_conn_id": self.aws_conn_id,
+                "region": self.region,
+            },
+        )
+
+    async def run(self) -> AsyncIterator[TriggerEvent]:
+        async with EcsHook(aws_conn_id=self.aws_conn_id, region_name=self.region).async_conn as client:
+            waiter = client.get_waiter(self.waiter_name)
+            while self.attempts >= 1:
+                self.attempts = self.attempts - 1
+                try:
+                    await waiter.wait(
+                        clusters=[self.cluster_arn],
+                        WaiterConfig={
+                            "MaxAttempts": 1,
+                        },
+                    )
+                    # we reach this point only if the waiter met a success criteria
+                    yield TriggerEvent({"status": "success", "arn": self.cluster_arn})
+                    return
```

Comment:
I feel that, in this case, by using a generic Trigger for both the cluster_active and cluster_inactive, we are simplifying the code, but at the expense of user experience. Specifically, I think the exceptions raised should tell the user what the specific error was (i.e. create cluster failed etc.)

I don't know if this is a moot point because of the EMR Serverless custom waiters [PR](https://github.com/apache/airflow/pull/30463), which will clean up a lot of this code.

Reply:
I'd vote for not modifying that code now since we can replace it with the common helper as soon as it's merged